datajunction-server/datajunction_server/internal/nodes.py

```diff
@@ -813,15 +823,170 @@ async def update_cube_node(  # pylint: disable=too-many-locals
+        cube_dimensions = current_node_revision.cube_dimensions()
+        errors = []
+        try:
+            await validate_cube(
+                session,
+                metric_names=cube_metrics,
+                dimension_names=cube_dimensions,
+                require_dimensions=True,
+            )
+            current_node_revision.status = NodeStatus.VALID
+        except DJException as exc:  # pragma: no cover
+            current_node_revision.status = NodeStatus.INVALID
+            if exc.errors:
+                errors.extend(exc.errors)
+            else:
+                errors.append(
+                    DJError(code=ErrorCode.INVALID_DIMENSION, message=exc.message),
+                )
+        session.add(current_node_revision)
+        await session.commit()
+        return NodeValidator(
+            status=current_node_revision.status,
+            columns=current_node_revision.columns,
+            errors=errors,
+        )
+
+    # Revalidate all other node types
+    node_validator = await validate_node_data(current_node_revision, session)
+
+    # Update the status
+    node.current.status = node_validator.status  # type: ignore
+
+    # Check if any columns have been updated
+    existing_columns = {col.name: col for col in node.current.columns}  # type: ignore
+    updated_columns = False
+    for col in node_validator.columns:
+        if existing_col := existing_columns.get(col.name):
+            if existing_col.type != col.type:
+                existing_col.type = col.type
+                updated_columns = True
+        else:
+            node.current.columns.append(col)  # type: ignore
+            updated_columns = True
+
+    # Only create a new revision if the columns have been updated
+    if updated_columns:  # type: ignore
+        new_revision = copy_existing_node_revision(node.current)  # type: ignore
+        new_revision.version = str(
+            Version.parse(node.current.version).next_major_version(),  # type: ignore
+        )
+
+        new_revision.status = node_validator.status
+        new_revision.lineage = [
+            lineage.dict()
+            for lineage in await get_column_level_lineage(session, new_revision)
+        ]
+
+        # Save which columns were modified and update the columns with the changes
+        node_validator.updated_columns = node_validator.modified_columns(
+            new_revision,  # type: ignore
+        )
+        new_revision.columns = node_validator.columns
+
+        # Save the new revision of the child
+        node.current_version = new_revision.version  # type: ignore
+        new_revision.node_id = node.id  # type: ignore
+        session.add(node)
+        session.add(new_revision)
+    await session.commit()
+    await session.refresh(node.current)  # type: ignore
+    await session.refresh(node, ["current"])
+    return node_validator
+
+
+def topological_sort(nodes: List[Node]) -> List[Node]:
```

Comment:
This was the reason why the node update propagation wasn't working correctly: we need to sort the downstream dependencies in topological order, or it's possible that we'll update a child node before a parent node.